lib/Analysis/Liveness.cpp

```diff
@@ -0,0 +1,379 @@
+      if (getLiveness(successor)->isLiveIn(value) &&
+          visited.insert(successor).second)
+        toProcess.push_back(successor);
+    }
+  }
+
+  return result;
+}
+
+/// Gets liveness info (if any) for the block.
+const LivenessBlockInfo *Liveness::getLiveness(Block *block) const {
+  auto it = blockMapping.find(block);
+  return it == blockMapping.end() ? nullptr : &it->second;
+}
+
+/// Returns a reference to a set containing live-in values.
+const Liveness::ValueSetT &Liveness::getLiveIn(Block *block) const {
+  return getLiveness(block)->in();
+}
+
+/// Returns a reference to a set containing live-out values.
+const Liveness::ValueSetT &Liveness::getLiveOut(Block *block) const {
+  return getLiveness(block)->out();
+}
+
+/// Returns true if the given operation represent the last use of the
+/// given value.
+bool Liveness::isLastUse(Value *value, Operation *operation) const {
+  Block *block = operation->getBlock();
+  const LivenessBlockInfo *blockInfo = getLiveness(block);
+
+  // The given value escapes the associated block.
+  if (blockInfo->isLiveOut(value))
+    return false;
+
+  Operation *endOperation = blockInfo->getEndOperation(value, operation);
+  // If the operation is a real user of `value` the first check is sufficient.
+  // If not, we will have to test whether the end operation is executed before
+  // the given operation in the block.
+  return endOperation == operation || endOperation->isBeforeInBlock(operation);
+}
+
+/// Dumps the liveness information in a human readable format.
+void Liveness::dump() const { print(llvm::errs()); }
+
+/// Dumps the liveness information to the given stream.
+void Liveness::print(raw_ostream &os) const {
+  os << "// ---- Liveness -----\n";
+
+  // Builds unique block/value mappings for testing purposes.
+  llvm::DenseMap<Block *, size_t> blockIds;
+  llvm::DenseMap<Operation *, size_t> operationIds;
+  llvm::DenseMap<Value *, size_t> valueIds;
+  for (Region &region : operation->getRegions())
+    for (Block &block : region) {
+      blockIds[&block] = blockIds.size();
+      for (BlockArgument *argument : block.getArguments())
+        valueIds[argument] = valueIds.size();
+      for (Operation &operation : block) {
+        operationIds[&operation] = operationIds.size();
+        for (Value *result : operation.getResults())
+          valueIds[result] = valueIds.size();
+      }
+    }
+
+  // Local printing helpers
+  auto printValueRef = [&](Value *value) {
+    if (Operation *defOp = value->getDefiningOp())
+      os << "val_" << defOp->getName();
+    else {
+      auto blockArg = cast<BlockArgument>(value);
+      os << "arg" << blockArg->getArgNumber() << "@"
+         << blockIds[blockArg->getOwner()];
+    }
+    os << " ";
+  };
+
+  auto printValueRefs = [&](const ValueSetT &values) {
+    std::vector<Value *> orderedValues(values.begin(), values.end());
+    std::sort(orderedValues.begin(), orderedValues.end(),
```

Comment:
Use llvm::array_pod_sort instead.

Reply:
Unfortunately, the `llvm::array_pod_sort` function does not seem to be compatible with a local lambda closure at first sight. However, this is a prerequisite in this case, since we access internal data structures for comparisons. According to the documentation of `llvm::array_pod_sort` we should use `std::sort` in this case. What do you think?